diff --git a/modules/dcache-bulk/src/main/java/org/dcache/services/bulk/BulkService.java b/modules/dcache-bulk/src/main/java/org/dcache/services/bulk/BulkService.java
new file mode 100644
index 0000000..01194a3
--- /dev/null
+++ b/modules/dcache-bulk/src/main/java/org/dcache/services/bulk/BulkService.java
@@ -0,0 +1,359 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Required;
+
+import javax.security.auth.Subject;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+import diskCacheV111.vehicles.Message;
+
+import dmg.cells.nucleus.CellLifeCycleAware;
+import dmg.cells.nucleus.CellMessageReceiver;
+import dmg.cells.nucleus.Reply;
+
+import org.dcache.auth.attributes.Restriction;
+import org.dcache.auth.attributes.Restrictions;
+import org.dcache.cells.MessageReply;
+import org.dcache.services.bulk.handlers.BulkSubmissionHandler;
+import org.dcache.services.bulk.queue.BulkServiceQueue;
+import org.dcache.services.bulk.store.BulkRequestStore;
+import org.dcache.services.bulk.util.BulkServiceStatistics;
+
+/**
+ *  Handles incoming messages.  Handles restart reloading.
+ *  Provides components for request and job processing.
+ */
+public class BulkService implements CellLifeCycleAware, CellMessageReceiver
+{
+    private static final Logger           LOGGER
+                    = LoggerFactory.getLogger(BulkService.class);
+
+    private BulkRequestStore      requestStore;
+    private BulkServiceQueue      queue;
+    private BulkSubmissionHandler submissionHandler;
+
+    private ExecutorService       incomingExecutorService;
+    private BulkServiceStatistics statistics;
+    private int                   maxRequestsPerUser;
+
+    @Override
+    public void afterStart()
+    {
+        /*
+         * Maintaining checkpoints on jobs that have been launched
+         * is currently unimplemented (it presents several notable
+         * complications).
+         *
+         * Since individual jobs are thus considered lost on restart, we
+         * must treat all unfinished requests as if they were idempotent.
+         *
+         * We thus reload the requests from the persistent store into memory,
+         * resetting any requests with non-terminal states to QUEUED
+         * (the latter is taken care of by the underlying implementation.)
+         *
+         * There is no danger in a race here because the state of the
+         * requests is not checked until the job queue is initialized
+         * and started.
+         */
+        try {
+            LOGGER.info("Loading requests into the request store/queue; "
+                                        + "incomplete requests will be "
+                                        + "reset to QUEUED.");
+            requestStore.load();
+        } catch (BulkServiceException e) {
+            LOGGER.error("There was a problem reloading requests {}.",
+                         e.toString());
+        }
+
+        LOGGER.info("Initializing the job queue.");
+        queue.initialize();
+
+        LOGGER.info("Signalling the job queue.");
+        queue.signal();
+
+        LOGGER.info("Service startup completed.");
+    }
+
+        @Override
+    public void beforeStop()
+    {
+        queue.shutdown();
+        try {
+            requestStore.save();
+        } catch (BulkServiceException e) {
+            LOGGER.warn("Problem storing on shutdown: {}.", e.toString());
+        }
+    }
+
+    public Reply messageArrived(BulkRequestMessage message)
+    {
+        LOGGER.trace("received BulkRequestMessage {}", message);
+
+        MessageReply<Message> reply = new MessageReply<>();
+        incomingExecutorService.execute(() -> {
+            try {
+                BulkRequest request = message.getRequest();
+                Subject subject = message.getSubject();
+                checkQuota(subject);
+                checkRestrictions(message.getRestriction(), request.getId());
+                requestStore.store(subject,
+                                   message.getRestriction(),
+                                   request,
+                                   null);
+                queue.signal();
+                statistics.incrementRequestsReceived(request.getActivity());
+                reply.reply(message);
+            } catch (BulkServiceException e) {
+                LOGGER.error("messageArrived(BulkRequestMessage) {}: {}",
+                             message, e.toString());
+                reply.fail(message, e);
+            }
+        });
+        return reply;
+    }
+
+    public Reply messageArrived(BulkRequestListMessage message)
+    {
+        LOGGER.trace("received BulkRequestListMessage {}", message);
+
+        MessageReply<Message> reply = new MessageReply<>();
+        incomingExecutorService.execute(() -> {
+            try {
+                List<String> requests
+                    = requestStore.getRequestUrls(message.getSubject(),
+                                                  message.getStatus());
+                message.setRequests(requests);
+                reply.reply(message);
+            } catch (BulkServiceException e) {
+                LOGGER.error("messageArrived(BulkRequestListMessage) {}: {}",
+                             message, e.toString());
+                reply.fail(message, e);
+            }
+        });
+        return reply;
+    }
+
+    public Reply messageArrived(BulkRequestStatusMessage message)
+    {
+        LOGGER.trace("received BulkRequestStatusMessage {}", message);
+
+        MessageReply<Message> reply = new MessageReply<>();
+        incomingExecutorService.execute(() -> {
+            try {
+                String requestId = message.getRequestId();
+                checkRestrictions(message.getRestriction(), requestId);
+                /*
+                 *  Checks permissions and presence of request.
+                 */
+                BulkRequestStatus status
+                                = requestStore.getStatus(message.getSubject(),
+                                                         requestId);
+                message.setStatus(status);
+                reply.reply(message);
+            } catch (BulkServiceException e) {
+                LOGGER.error("messageArrived(BulkRequestStatusMessage) {}: {}",
+                             message, e.toString());
+                reply.fail(message, e);
+            }
+        });
+        return reply;
+    }
+
+    public Reply messageArrived(BulkRequestCancelMessage message)
+    {
+        LOGGER.trace("received BulkRequestCancelMessage {}", message);
+
+        MessageReply<Message> reply = new MessageReply<>();
+        incomingExecutorService.execute(() -> {
+            try {
+                String requestId = message.getRequestId();
+                checkRestrictions(message.getRestriction(), requestId);
+                /*
+                 *  Checks permissions and presence of request.
+                 */
+                submissionHandler.cancelRequest(message.getSubject(),
+                                                requestId);
+                reply.reply(message);
+            } catch (BulkServiceException e) {
+                LOGGER.error("messageArrived(BulkRequestCancelMessage) {}: {}",
+                             message, e.toString());
+                reply.fail(message, e);
+            }
+        });
+        return reply;
+    }
+
+    public Reply messageArrived(BulkRequestClearMessage message)
+    {
+        LOGGER.trace("received BulkRequestClearMessage {}", message);
+
+        MessageReply<Message> reply = new MessageReply<>();
+        incomingExecutorService.execute(() -> {
+            try {
+                String requestId = message.getRequestId();
+                checkRestrictions(message.getRestriction(), requestId);
+                /*
+                 *  Checks permissions and presence of request.
+                 */
+                submissionHandler.clearRequest(message.getSubject(),
+                                               requestId);
+                reply.reply(message);
+            } catch (BulkServiceException e) {
+                LOGGER.error("messageArrived(BulkRequestClearMessage) {}: {}",
+                             message, e.toString());
+                reply.fail(message, e);
+            }
+        });
+        return reply;
+    }
+
+    public BulkServiceQueue getQueue()
+    {
+        return queue;
+    }
+
+    public BulkSubmissionHandler getSubmissionHandler()
+    {
+        return submissionHandler;
+    }
+
+    public BulkRequestStore getRequestStore()
+    {
+        return requestStore;
+    }
+
+    @Required
+    public void setIncomingExecutorService(ExecutorService incomingExecutorService)
+    {
+        this.incomingExecutorService = incomingExecutorService;
+    }
+
+    @Required
+    public void setMaxRequestsPerUser(int maxRequestsPerUser)
+    {
+        this.maxRequestsPerUser = maxRequestsPerUser;
+    }
+
+    @Required
+    public void setQueue(BulkServiceQueue queue)
+    {
+        this.queue = queue;
+    }
+
+    @Required
+    public void setSubmissionHandler(BulkSubmissionHandler submissionHandler)
+    {
+        this.submissionHandler = submissionHandler;
+    }
+
+    @Required
+    public void setRequestStore(BulkRequestStore requestStore)
+    {
+        this.requestStore = requestStore;
+    }
+
+    @Required
+    public void setStatistics(BulkServiceStatistics statistics)
+    {
+        this.statistics = statistics;
+    }
+
+    /**
+     * This may be subject to change.
+     *
+     * @param restriction from the incoming message
+     * @param requestId
+     * @throws BulkPermissionDeniedException
+     * @throws BulkRequestStorageException
+     */
+    private void checkRestrictions(Restriction restriction,
+                                   String requestId)
+                    throws BulkPermissionDeniedException,
+                    BulkRequestStorageException
+    {
+        Optional<Restriction> option = requestStore.getRestriction(requestId);
+        Restriction original = option.orElse(null);
+
+        if (restriction == null || Restrictions.none().equals(restriction)) {
+            return;
+        }
+
+        if (original != null && !original.isSubsumedBy(restriction)) {
+            throw new BulkPermissionDeniedException(requestId);
+        }
+    }
+
+    private void checkQuota(Subject subject) throws BulkQuotaExceededException,
+                    BulkRequestStorageException
+    {
+        String user = BulkRequestStore.uidGidKey(subject);
+        if (requestStore.countNonTerminated(user) >= maxRequestsPerUser) {
+            throw new BulkQuotaExceededException(user);
+        }
+        statistics.addUserRequest(user);
+    }
+}
diff --git a/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkQuotaExceededException.java b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkQuotaExceededException.java
new file mode 100644
index 0000000..4448710
--- /dev/null
+++ b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkQuotaExceededException.java
@@ -0,0 +1,75 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+public class BulkQuotaExceededException extends BulkServiceException
+{
+    private static final long serialVersionUID = -2776212873761086957L;
+
+    public BulkQuotaExceededException(String message)
+    {
+        super(message);
+    }
+
+    public BulkQuotaExceededException(String message, Throwable cause)
+    {
+        super(message, cause);
+    }
+}
diff --git a/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestCancelMessage.java b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestCancelMessage.java
new file mode 100644
index 0000000..8ec629e
--- /dev/null
+++ b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestCancelMessage.java
@@ -0,0 +1,82 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+import org.dcache.auth.attributes.Restriction;
+
+/**
+ *   Message to cancel a single request.
+ */
+public class BulkRequestCancelMessage extends BulkServiceMessage
+{
+    private static final long serialVersionUID = -7362655973765754365L;
+    private final String requestId;
+
+    public BulkRequestCancelMessage(String requestId, Restriction restriction)
+    {
+        super(restriction);
+        this.requestId = requestId;
+    }
+
+    public String getRequestId()
+    {
+        return requestId;
+    }
+}
diff --git a/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestClearMessage.java b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestClearMessage.java
new file mode 100644
index 0000000..787b08f
--- /dev/null
+++ b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestClearMessage.java
@@ -0,0 +1,82 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+import org.dcache.auth.attributes.Restriction;
+
+/**
+ *   Message to clear a single request.
+ */
+public class BulkRequestClearMessage extends BulkServiceMessage
+{
+    private static final long serialVersionUID = 957050116233062328L;
+    private final String       requestId;
+
+    public BulkRequestClearMessage(String requestId, Restriction restriction)
+    {
+        super(restriction);
+        this.requestId = requestId;
+    }
+
+    public String getRequestId()
+    {
+        return requestId;
+    }
+}
diff --git a/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestListMessage.java b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestListMessage.java
new file mode 100644
index 0000000..f28c5c2
--- /dev/null
+++ b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestListMessage.java
@@ -0,0 +1,107 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+import org.dcache.auth.attributes.Restriction;
+import org.dcache.services.bulk.BulkRequestStatus.Status;
+
+/**
+ *   To acquire a list of requests submitted by the user.  May optionally
+ *   be filtered by status.
+ */
+public class BulkRequestListMessage extends BulkServiceMessage
+{
+    private static final long serialVersionUID = -3921177984889456203L;
+
+    private final Set<Status>  status;
+    private       List<String> requests;
+
+    public BulkRequestListMessage(Restriction restriction)
+    {
+        this(null, restriction);
+    }
+
+    public BulkRequestListMessage(Set<Status> status, Restriction restriction)
+    {
+        super(restriction);
+        this.status = status == null ? Collections.emptySet() : status;
+    }
+
+    public Set<Status> getStatus()
+    {
+        return status;
+    }
+
+    public List<String> getRequests()
+    {
+        return requests;
+    }
+
+    public void setRequests(List<String> requests)
+    {
+        this.requests = requests;
+    }
+}
+
+
diff --git a/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestMessage.java b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestMessage.java
new file mode 100644
index 0000000..bdf8645
--- /dev/null
+++ b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestMessage.java
@@ -0,0 +1,91 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+import org.dcache.auth.attributes.Restriction;
+
+/**
+ *   Single generic bulk message type which serves various activities.
+ */
+public class BulkRequestMessage extends BulkServiceMessage
+{
+    private static final long serialVersionUID = 957050116233062328L;
+
+    private final BulkRequest request;
+
+    public BulkRequestMessage(BulkRequest request, Restriction restriction)
+    {
+        super(restriction);
+        this.request = request;
+    }
+
+    public BulkRequest getRequest()
+    {
+        return request;
+    }
+
+    public String toString()
+    {
+        return "BulkRequestMessage for request " + request.getUrlPrefix()
+                        + "/" + request.getId() + ", target: "
+                        + request.getTarget() + ", activity: "
+                        + request.getActivity();
+    }
+}
diff --git a/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestStatusMessage.java b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestStatusMessage.java
new file mode 100644
index 0000000..ec2c51b
--- /dev/null
+++ b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkRequestStatusMessage.java
@@ -0,0 +1,94 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+import org.dcache.auth.attributes.Restriction;
+
+/**
+ * Single generic bulk message type which serves various activities.
+ */
+public class BulkRequestStatusMessage extends BulkServiceMessage
+{
+    private static final long serialVersionUID = 3470455902659932131L;
+
+    private final String            requestId;
+    private       BulkRequestStatus status;
+
+    public BulkRequestStatusMessage(String requestId, Restriction restriction)
+    {
+        super(restriction);
+        this.requestId = requestId;
+    }
+
+    public String getRequestId()
+    {
+        return requestId;
+    }
+
+    public BulkRequestStatus getStatus()
+    {
+        return status;
+    }
+
+    public void setStatus(BulkRequestStatus status)
+    {
+        this.status = status;
+    }
+}
diff --git a/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkServiceMessage.java b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkServiceMessage.java
new file mode 100644
index 0000000..d96543f
--- /dev/null
+++ b/modules/dcache-vehicles/src/main/java/org/dcache/services/bulk/BulkServiceMessage.java
@@ -0,0 +1,79 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.services.bulk;
+
+import diskCacheV111.vehicles.Message;
+
+import org.dcache.auth.attributes.Restriction;
+
+public abstract class BulkServiceMessage extends Message
+{
+    protected final Restriction restriction;
+
+    protected BulkServiceMessage(Restriction restriction)
+    {
+        this.restriction = restriction;
+    }
+
+    public Restriction getRestriction()
+    {
+        return restriction;
+    }
+}
