diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/data/PoolQoSStatus.java b/modules/dcache-qos/src/main/java/org/dcache/qos/data/PoolQoSStatus.java
new file mode 100644
index 0000000000..f674c288e0
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/data/PoolQoSStatus.java
@@ -0,0 +1,129 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.data;
+
+import diskCacheV111.pools.PoolV2Mode;
+
+/**
+ *  Interprets a {@link PoolV2Mode} for potential qos handling.
+ *  <p/>
+ *  Shared between verifier and scanner.
+ */
+public enum PoolQoSStatus {
+    UNINITIALIZED,      // no info yet
+    DOWN,               // equivalent to the client's being unable to read from pool
+    ENABLED,            // normal read-write operations are possible
+    READ_ONLY;          // equivalent to disabled for writing by clients
+
+    /**
+     *  This status tells qos whether action (scanning) needs to
+     *  be taken with respect to the pool.  It is
+     *  slightly different from the pool's mode.
+     *
+     *  @return the status equivalent to the mode.
+     */
+    public static PoolQoSStatus getStatusFor(PoolV2Mode poolMode) {
+        if (poolMode == null) {
+            return UNINITIALIZED;
+        }
+
+        /*
+         *  It does not matter whether we can still write to this pool;
+         *  for the purposes of qos, if the pool is not readable
+         *  by the client, it should be treated as down.
+         */
+        if (poolMode.getMode() == PoolV2Mode.DISABLED ||
+                        poolMode.isDisabled(PoolV2Mode.DISABLED_DEAD) ||
+                        poolMode.isDisabled(PoolV2Mode.DISABLED_FETCH)) {
+            return DOWN;
+        }
+
+        /*
+         *  For qos, 'READ_ONLY' should designate only the fact
+         *  that clients cannot write to the pool, or that staging
+         *  is turned off.
+         *
+         *  Whether p2p_client is enabled or disabled does not matter here;
+         *  it is reflected in the selection algorithm, but otherwise it
+         *  does not constitute a reason to migrate existing replicas.
+         */
+        if (poolMode.isDisabled(PoolV2Mode.DISABLED_STORE) ||
+                        poolMode.isDisabled(PoolV2Mode.DISABLED_STAGE)) {
+            return READ_ONLY;
+        }
+
+        /*
+         *  Not exactly the same as PoolV2Mode.ENABLED.  This
+         *  signifies that the pool can be written to and read from
+         *  by external clients or is available for staging.
+         */
+        return ENABLED;
+    }
+
+    public QoSMessageType getMessageType() {
+        switch (this) {
+            case DOWN:
+            case UNINITIALIZED:
+                return QoSMessageType.POOL_STATUS_DOWN;
+            default:
+                return QoSMessageType.POOL_STATUS_UP;
+        }
+    }
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSActionCompletedListener.java b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSActionCompletedListener.java
new file mode 100644
index 0000000000..ab01518318
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSActionCompletedListener.java
@@ -0,0 +1,76 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.listeners;
+
+import diskCacheV111.util.PnfsId;
+import java.io.Serializable;
+import org.dcache.qos.data.QoSAction;
+
+public interface QoSActionCompletedListener {
+  /**
+   *  Notification that a single adjustment activity for a file has completed.
+   *
+   *  @param pnfsId of the file.
+   *  @param action copying, changing the sticky bit, flushing or staging the file.
+   *  @param error  if the action failed (this will be the last error associated with the
+   *                verification operation).
+   */
+  void fileQoSActionCompleted(PnfsId pnfsId, QoSAction action, Serializable error);
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSAdjustmentListener.java b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSAdjustmentListener.java
new file mode 100644
index 0000000000..ac7456e992
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSAdjustmentListener.java
@@ -0,0 +1,80 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.listeners;
+
+import diskCacheV111.util.PnfsId;
+import org.dcache.qos.QoSException;
+import org.dcache.qos.vehicles.QoSAdjustmentRequest;
+
+public interface QoSAdjustmentListener {
+  /**
+   *  Notification that something needs to be done to satisfy the file's QoS requirements.
+   *
+   *  @param adjustmentRequest indicating file, metadata and the action which should be taken.
+   */
+  void fileQoSAdjustmentRequested(QoSAdjustmentRequest adjustmentRequest) throws QoSException;
+
+  /**
+   *  Notify that the current adjustment should be canceled for the given file.
+   *
+   *  @param pnfsId file for which to cancel any adjustment in flight.
+   */
+  void fileQoSAdjustmentCancelled(PnfsId pnfsId) throws QoSException;
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSPoolScanResponseListener.java b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSPoolScanResponseListener.java
new file mode 100644
index 0000000000..baee88236c
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSPoolScanResponseListener.java
@@ -0,0 +1,74 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.listeners;
+
+public interface QoSPoolScanResponseListener {
+    /**
+     *   Report completed adjustments or verifications for files residing on a pool.
+     *   Usually successive such updates are to be accumulated by the receiver (they represent
+     *   batches of work). The idea is to give the receiver an indication of how much work has
+     *   been done up to this point.
+     *
+     *   @param pool on which the scan is being done.
+     *   @param succeeded number of requests that were successful; can be a partial count.
+     *   @param failed number of requests that failed; can be a partial count.
+     */
+    void scanRequestUpdated(String pool, int succeeded, int failed);
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSVerificationListener.java b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSVerificationListener.java
new file mode 100644
index 0000000000..1f4696fce4
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/listeners/QoSVerificationListener.java
@@ -0,0 +1,114 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.listeners;
+
+import diskCacheV111.util.PnfsId;
+import org.dcache.qos.QoSException;
+import org.dcache.qos.vehicles.QoSAdjustmentResponse;
+import org.dcache.qos.vehicles.QoSBatchedVerificationRequest;
+import org.dcache.qos.vehicles.QoSVerificationRequest;
+
+public interface QoSVerificationListener {
+  /**
+   *  Request to check the file's status on disk and tape against its requirements.
+   *
+   *  @param verificationRequest composing file update and requirements data.
+   */
+  void fileQoSVerificationRequested(QoSVerificationRequest verificationRequest) throws QoSException;
+
+  /**
+   *  Request to check a collection of files residing at a given (pool) location.
+   *
+   *  @param verificationRequest includes the location name, originating message type, and list of
+   *                             pnfsIds.
+   */
+  void fileQoSVerificationRequested(QoSBatchedVerificationRequest verificationRequest) throws QoSException;
+
+  /**
+   *  A single adjustment request has completed.
+   *
+   *  @param adjustmentResponse encapsulates the file id, action, and status of the adjustment.
+   */
+  void fileQoSAdjustmentCompleted(QoSAdjustmentResponse adjustmentResponse) throws QoSException;
+
+  /**
+   *  This is a notification to cancel and remove the verification operation for a file.
+   *
+   *  @param pnfsId the file for which to cancel verification operation.
+   */
+  void fileQoSVerificationCancelled(PnfsId pnfsId) throws QoSException;
+
+  /**
+   *  Scanning activity has been cancelled; this is a notification to cancel all outstanding
+   *  verification operations having this pool as 'parent'.
+   *
+   *  @param pool the parent pool to search for in order to cancel verification operations.
+   */
+  void fileQoSBatchedVerificationCancelled(String pool) throws QoSException;
+
+  /**
+   *  The verifier will usually need to know if a location has been manually excluded.
+   *  This information is conveyed by the scanner.
+   *
+   *  @param location whose exclusion status is being reported.
+   *  @param excluded whether it has been excluded (true) or re-included (false).
+   */
+  void notifyLocationExclusion(String location, boolean excluded) throws QoSException;
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSActionCompletedClient.java b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSActionCompletedClient.java
new file mode 100644
index 0000000000..593e5e3ee9
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSActionCompletedClient.java
@@ -0,0 +1,83 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.remote.clients;
+
+import diskCacheV111.util.PnfsId;
+import java.io.Serializable;
+import org.dcache.cells.CellStub;
+import org.dcache.qos.data.QoSAction;
+import org.dcache.qos.listeners.QoSActionCompletedListener;
+import org.dcache.vehicles.qos.QoSActionCompleteMessage;
+
+/**
+ *  Use this client when communicating with a remote requirements engine.
+ */
+public final class RemoteQoSActionCompletedClient implements QoSActionCompletedListener {
+  private CellStub requirementsService;
+
+  @Override
+  public void fileQoSActionCompleted(PnfsId pnfsId, QoSAction action, Serializable error) {
+    requirementsService.send(new QoSActionCompleteMessage(pnfsId, action, error));
+  }
+
+  public void setRequirementsService(CellStub requirementsService) {
+    this.requirementsService = requirementsService;
+  }
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSAdjustmentClient.java b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSAdjustmentClient.java
new file mode 100644
index 0000000000..45e935a345
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSAdjustmentClient.java
@@ -0,0 +1,89 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.remote.clients;
+
+import diskCacheV111.util.PnfsId;
+import org.dcache.cells.CellStub;
+import org.dcache.qos.QoSException;
+import org.dcache.qos.listeners.QoSAdjustmentListener;
+import org.dcache.qos.vehicles.QoSAdjustmentCancelledMessage;
+import org.dcache.qos.vehicles.QoSAdjustmentRequest;
+import org.dcache.qos.vehicles.QoSAdjustmentRequestMessage;
+
+/**
+ *  Use this client when communicating with a remote adjustment service.<
+ */
+public final class RemoteQoSAdjustmentClient implements QoSAdjustmentListener {
+  private CellStub adjustmentService;
+
+  @Override
+  public void fileQoSAdjustmentRequested(QoSAdjustmentRequest adjustmentRequest) {
+      adjustmentService.send(new QoSAdjustmentRequestMessage(adjustmentRequest));
+  }
+
+  @Override
+  public void fileQoSAdjustmentCancelled(PnfsId pnfsId) throws QoSException {
+    adjustmentService.send(new QoSAdjustmentCancelledMessage(pnfsId));
+  }
+
+  public void setAdjustmentService(CellStub adjustmentService) {
+    this.adjustmentService = adjustmentService;
+  }
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSScannerClient.java b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSScannerClient.java
new file mode 100644
index 0000000000..fdf0fc7218
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSScannerClient.java
@@ -0,0 +1,80 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.remote.clients;
+
+import org.dcache.cells.CellStub;
+import org.dcache.qos.listeners.QoSPoolScanResponseListener;
+import org.dcache.qos.vehicles.QoSBatchedVerificationResponseMessage;
+
+/**
+ *  Use this client when communicating with a remote scanner service.
+ */
+public final class RemoteQoSScannerClient implements QoSPoolScanResponseListener {
+  private CellStub scannerService;
+
+  @Override
+  public void scanRequestUpdated(String pool, int succeeded, int failed) {
+    scannerService.send(new QoSBatchedVerificationResponseMessage(pool, succeeded, failed));
+  }
+
+  public void setScannerService(CellStub scannerService) {
+    this.scannerService = scannerService;
+  }
+}
diff --git a/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSVerificationClient.java b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSVerificationClient.java
new file mode 100644
index 0000000000..ffc6cffd23
--- /dev/null
+++ b/modules/dcache-qos/src/main/java/org/dcache/qos/remote/clients/RemoteQoSVerificationClient.java
@@ -0,0 +1,114 @@
+/*
+COPYRIGHT STATUS:
+Dec 1st 2001, Fermi National Accelerator Laboratory (FNAL) documents and
+software are sponsored by the U.S. Department of Energy under Contract No.
+DE-AC02-76CH03000. Therefore, the U.S. Government retains a  world-wide
+non-exclusive, royalty-free license to publish or reproduce these documents
+and software for U.S. Government purposes.  All documents and software
+available from this server are protected under the U.S. and Foreign
+Copyright Laws, and FNAL reserves all rights.
+
+Distribution of the software available from this server is free of
+charge subject to the user following the terms of the Fermitools
+Software Legal Information.
+
+Redistribution and/or modification of the software shall be accompanied
+by the Fermitools Software Legal Information  (including the copyright
+notice).
+
+The user is asked to feed back problems, benefits, and/or suggestions
+about the software to the Fermilab Software Providers.
+
+Neither the name of Fermilab, the  URA, nor the names of the contributors
+may be used to endorse or promote products derived from this software
+without specific prior written permission.
+
+DISCLAIMER OF LIABILITY (BSD):
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED  WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED  WARRANTIES OF MERCHANTABILITY AND FITNESS
+FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL FERMILAB,
+OR THE URA, OR THE U.S. DEPARTMENT of ENERGY, OR CONTRIBUTORS BE LIABLE
+FOR  ANY  DIRECT, INDIRECT,  INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY  OF
+LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT  OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE  POSSIBILITY OF SUCH DAMAGE.
+
+Liabilities of the Government:
+
+This software is provided by URA, independent from its Prime Contract
+with the U.S. Department of Energy. URA is acting independently from
+the Government and in its own private capacity and is not acting on
+behalf of the U.S. Government, nor as its contractor nor its agent.
+Correspondingly, it is understood and agreed that the U.S. Government
+has no connection to this software and in no manner whatsoever shall
+be liable for nor assume any responsibility or obligation for any claim,
+cost, or damages arising out of or resulting from the use of the software
+available from this server.
+
+Export Control:
+
+All documents and software available from this server are subject to U.S.
+export control laws.  Anyone downloading information from this server is
+obligated to secure any necessary Government licenses before exporting
+documents or software obtained from this server.
+ */
+package org.dcache.qos.remote.clients;
+
+import diskCacheV111.util.PnfsId;
+import org.dcache.cells.CellStub;
+import org.dcache.qos.QoSException;
+import org.dcache.qos.listeners.QoSVerificationListener;
+import org.dcache.qos.vehicles.QoSAdjustmentResponse;
+import org.dcache.qos.vehicles.QoSAdjustmentResponseMessage;
+import org.dcache.qos.vehicles.QoSBatchVerificationCancelledMessage;
+import org.dcache.qos.vehicles.QoSBatchedVerificationRequest;
+import org.dcache.qos.vehicles.QoSBatchedVerificationRequestMessage;
+import org.dcache.qos.vehicles.QoSLocationExcludedMessage;
+import org.dcache.qos.vehicles.QoSVerificationCancelledMessage;
+import org.dcache.qos.vehicles.QoSVerificationRequest;
+import org.dcache.qos.vehicles.QoSVerificationRequestMessage;
+
+/**
+ *  Use this client when communicating with a remote verification service.
+ */
+public final class RemoteQoSVerificationClient implements QoSVerificationListener {
+  private CellStub verificationService;
+
+  @Override
+  public void fileQoSVerificationRequested(QoSVerificationRequest verificationRequest) {
+    verificationService.send(new QoSVerificationRequestMessage(verificationRequest));
+  }
+
+  @Override
+  public void fileQoSVerificationRequested(QoSBatchedVerificationRequest verificationRequest) {
+    verificationService.send(new QoSBatchedVerificationRequestMessage(verificationRequest));
+  }
+
+  @Override
+  public void fileQoSAdjustmentCompleted(QoSAdjustmentResponse adjustmentResponse)
+      throws QoSException {
+    verificationService.send(new QoSAdjustmentResponseMessage(adjustmentResponse));
+  }
+
+  @Override
+  public void fileQoSVerificationCancelled(PnfsId pnfsId) throws QoSException {
+    verificationService.send(new QoSVerificationCancelledMessage(pnfsId));
+  }
+
+  public void fileQoSBatchedVerificationCancelled(String pool) {
+    verificationService.send(new QoSBatchVerificationCancelledMessage(pool));
+  }
+
+  public void notifyLocationExclusion(String location, boolean excluded) {
+    verificationService.send(new QoSLocationExcludedMessage(location, excluded));
+  }
+
+  public void setVerificationService(CellStub verificationService) {
+    this.verificationService = verificationService;
+  }
+}
